race-buffer/model/SequenceNumbers.tla

```diff
@@ -0,0 +1,185 @@
+--------------------------- MODULE SequenceNumbers ---------------------------
+EXTENDS Naturals
+
+CONSTANTS 
+  IntModulus, (* Modulus of integers (maximum value + 1) *)
+
+  IncrementTo  (* The value of the sequence number that the writer should stop at *)
+-----------------------------------------------------------------------------
+
+VARIABLES 
+  wpc, (* Writer's program counter *)
+  written_low, (* Low word of sequence number written in memory *)
+  written_high, (* High word of sequence number written in memory *)
+
+  rpc, (* Reader's program counter *)
+  snapped_low, (* Snapshot of the low word *)
+  snapped_high, (* Snapshot of the high word *)
+  read_seqn, (* The last sequence number successfully read *)
+  (* The last recorded in-memory sequence number before the start of the last read
+     (Not actually known by reader, just used for invariants) *)
+  seqn_before_read
+
+-----------------------------------------------------------------------------
+
+(* Writer helpers *)
+
+(* Calculate the integer-valued sequence number based on the high and low words *)
+SeqNum(high, low) == (high.val * IntModulus) + low 
+
+(* Structure used to store a word and a boolean set to true when the value is getting edited.
+   In reality, the high bit of the high word is used instead of a separate boolean *)
+ProgressValStruct(val, in_progress) == [val |-> val, in_progress |-> in_progress]
+
+(* Create ProgressValStruct not in progress *)
+ProgressVal(val) == ProgressValStruct(val, FALSE)
+
+(* Set ProgressVal structure to be in progress *)
+SetInProgress(progress_struct) == ProgressValStruct(progress_struct.val, TRUE)
+
+-----------------------------------------------------------------------------
+
+(* Writer steps *)
+
+(* The low word of the writer is incremented until it reaches the integer maximum *)
+IncrementLow ==
```

Comment:
Note for reference that we should model incrementing both read and write by both 1 and 2 (due to being able to read 1 or 2 log entries).